feat: CSS Modules support (#4852)
remix-run/remix: docs/guides/styling.md

NOTE: You may run into hydration warnings when using Styled Components. Hopefully [this issue][styled-components-issue] will be fixed soon.

## CSS Modules

<docs-warning>This feature is unstable and currently only available behind a feature flag. We're confident in the use cases it solves but the API and implementation may change in the future.</docs-warning>

> <docs-warning>Since all CSS Modules styles end up in a single CSS file at the end of the build, we recommend that you only use this styling approach for smaller applications, or sparingly within larger applications. If you're concerned about bundle size, you should probably look at [Tailwind](#tailwind-css) or [Regular Stylesheets](#regular-stylesheets) instead.

Reviewer: I'm actually not sure that we need this warning. I think a single module bundle is probably fine for decently large apps since it's easy to cache, and CSS files are rarely a bottleneck before some pretty serious scale. I think we can talk about the tradeoffs without making it look scary.

Reviewer 2: It does sound a little scary if someone were evaluating remix and saw CSS modules
Not discounting that the warning isn't true. Maybe there are strategies for reducing the bundle size like minifying the CSS.

just noticed out production CSS output isn't compressed in our app 😅

Reviewer: > Remix doesn't minify CSS by default?

No, Remix doesn't process CSS by default at all (aside than hashing the filename for effective versioning/caching). This is explained in the same doc.

That said, CSS Modules is opening the door a bit on this bigger question. We should absolutely be able to minify the production build of your CSS Modules bundle IMO.

Reviewer 3: > I think a single module bundle is probably fine for decently large apps since it's easy to cache

But change any style will break the cache.

Reviewer: > But change any style will break the cache.

As it should! I'm not saying there are no potential perf tradeoffs here, just that we shouldn't overstate them because:

> CSS files are rarely a bottleneck before some pretty serious scale

Author: >We should absolutely be able to minify the production build of your CSS Modules bundle IMO.

The generated CSS is already being minified since it's being bundled by esbuild, honouring the same minification settings as the JS output.

</docs-warning>

To enable [CSS Modules], set the `future.unstable_cssModules` feature flag in `remix.config.js`.

```js filename=remix.config.js
/** @type {import('@remix-run/dev').AppConfig} */
module.exports = {
  future: {
    unstable_cssModules: true,
  },
  // ...
};
```

> With this feature flag enabled, the Remix compiler will now generate a single CSS file containing all CSS Modules styles in your application. Note that any [regular stylesheet imports](http://localhost:3000/docs/en/v1/guides/styling#regular-stylesheets) will remain as separate files.

Reviewer: ```suggestion
With this feature flag enabled, the Remix compiler will now generate a single CSS file containing all CSS Modules styles in your application. Note that any [regular stylesheet imports](#regular-stylesheets) will remain as separate files.
```

Author: Whoooooops.

> Unlike many other tools in the React ecosystem, insertion of this CSS bundle into the page is not handled for you automatically. Instead, we ensure that you always have control over the link tags on your page. This lets you decide where the CSS bundle is inserted and in what order relative to other style sheets.

Reviewer: ```suggestion
Unlike many other tools in the React ecosystem, we do not insert the CSS bundle into the page automatically. Instead, we ensure that you always have control over the link tags on your page. This lets you decide where the CSS file is loaded relative to other stylesheets in your app.
```

To get access to the CSS bundle, first install the `@remix-run/css-bundle` package.

```sh
npm install @remix-run/css-bundle
```

> Then, import `cssBundleHref` and add it to a link descriptor, most likely in `root.tsx` so that it applies to your entire application.

Reviewer: ```suggestion
Then, import `cssBundleHref` and add it to a link descriptor—most likely in `root.tsx` so that it applies to your entire application.
```

Author: Can't believe I passed up an opportunity to use an em dash.

```tsx filename=root.tsx lines=[2,6]
import type { LinksFunction } from "@remix-run/node"; // or cloudflare/deno

> import cssBundleHref from "@remix-run/css-bundle";

Reviewer: I'm wondering if we shouldn't use a more specific virtual here for CSS Modules. If we take a similar approach with raw CSS imports and Vanilla Extract, it could be nice to have access to each of those bundles independently (ex: your admin routes are styled w/ CSS Modules, marketing pages use some other bundle)

Reviewer: Perhaps we could achieve both?

```ts
// all bundles in one stylesheet
import cssBundleStylesheet from "@remix-run/css-bundle";

// split up
import cssModulesStylesheet from "@remix-run/css-bundle/css-modules";
import vanillaExtractStylesheet from "@remix-run/css-bundle/vanilla-extract";
```

Author: We don't get that level of granularity since we're now hooking into esbuild's CSS loader, which means we'll only get a single file at the end. I'm not convinced this is a big enough use case to warrant implementing our own CSS bundling logic, compared to bundle splitting of CSS for example.

It also sounds like a bit of a fragile pattern to me since nothing enforces that you're only using CSS Modules / Vanilla Extract / etc. within a given section of your application, leading to missing styles if you happen to use these features elsewhere.

Reviewer: Good points. Perhaps another thing we attempt to solve later by splitting CSS at the route level.

export const links: LinksFunction = () => {
  return [
    { rel: "stylesheet", href: cssBundleHref },
    // ...
  ];
};
```

> You're all set! You can now opt into CSS Modules via the `.module.css` file name convention, for example:

Reviewer: ```suggestion
You're all set! You can now opt into CSS Modules via the `.module.css` file name convention. For example:
```
